srcmedia/ts/lib/input.ts

```diff
@@ -1,115 +1,105 @@
+        )
     }
 }
 
 /**
- * A reactive <input type="text"> element.
+ * A reactive `<input type="text">` element. Stores its `value` as an observable
+ * `string`.
  *
  * @class RxTextInput
  * @extends {RxInput}
- * @implements {Reactive<RxTextInputState>}
  */
-class RxTextInput extends RxInput implements Reactive<RxTextInputState> {
-    state: Subject<RxTextInputState>
+class RxTextInput extends RxInput {
+
+    public value$: Observable<string>
 
     constructor(element: HTMLInputElement) {
         super(element)
-        // Update state when the user types in a new value, with debounce
-        fromEvent(this.element, 'input').pipe(
-            map(() => ({ value: this.element.value })),
-            debounceTime(500),
-            distinctUntilChanged()
-        ).subscribe(this.update)
+        this.value$ = this.events$.pipe(
+            map(() => this.element.value),
+            startWith(this.element.value),
+        )
     }
 }
 
 /**
- * A reactive <input type="checkbox"> element.
+ * A reactive `<input type="number">` element. Stores its `value` as an
+ * observable `number`. Non-integer values will be parsed to `NaN`.
  *
- * @class RxCheckboxInput
+ * @class RxNumberInput
  * @extends {RxInput}
- * @implements {Reactive<RxCheckboxInputState>}
  */
-class RxCheckboxInput extends RxInput implements Reactive<RxCheckboxInputState>{
-    state: Subject<RxCheckboxInputState>
+class RxNumberInput extends RxInput {
+
+    public value$: Observable<number>
 
     constructor(element: HTMLInputElement) {
         super(element)
-        // Update state immediately when the checkbox is clicked
-        fromEvent(this.element, 'change').pipe(
-            map(() => ({ checked: this.element.checked })),
-        ).subscribe(this.update)
+        this.value$ = this.events$.pipe(
+            map(() => this.element.value),
+            map(value => parseInt(value)), // will return NaN if not an integer
+            startWith(parseInt(this.element.value)),
+        )
     }
+}
 
-    /**
-     * Update state and propagate changes to the checkbox element.
-     *
-     * @param {RxCheckboxInputState} state
-     * @returns {Promise<void>}
-     * @memberof RxCheckboxInput
-     */
-    update = async (newState: Partial<RxCheckboxInputState>): Promise<void> => {
-        if (newState.value) this.element.value = newState.value
-        if (newState.checked) this.element.checked = newState.checked
-        this.state.next({
-            value: this.element.value,
-            checked: this.element.checked
-        })
+/**
+ * A reactive `<input type="checkbox">` element. Stores its checked state as an
+ * observable `boolean`.
+ *
+ * @class RxCheckboxInput
+ * @extends {RxInput}
+ */
+class RxCheckboxInput extends RxInput {
+
+    public checked$: Observable<boolean>
+
+    protected readonly EVENT_TYPE: string = 'change' // use the 'change' event for legacy compatibility
```

Comment:
does "legacy compatibility" mean specific browsers?

Reply:
iOS Safari for sure, and also I think IE. there's [some info here](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event#Browser_compatibility) but I think I read a more detailed analysis of the compatibility of it...can't find now. bottom line is that _all_ browsers should have _all_ `<inputs>` support the 'input' event type as per the spec, but some still don't, and some non-supported (IE) ones never will

Reply:
Worth adding that link to the comments?

Reply:
might as well, since it's a starting point for finding more info. thx